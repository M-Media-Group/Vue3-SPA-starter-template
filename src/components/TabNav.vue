<script setup lang="ts">
import type { PropType } from "vue";
import BaseButton from "./BaseButton.vue";
import type { selectOption } from "@/types/listItem";
import { useMultiselect } from "@/composables/useMultiselect";

const emit = defineEmits([
  /** The page the user has navigated to, either by clicking directly on a page or by using the previous and next buttons */
  "update:modelValue",
]);

const props = defineProps({
  /** THe modelValue, which is an array of page IDs that are currently active */
  modelValue: {
    type: Array as PropType<string[]>,
    required: false,
    default: () => [],
  },

  /** The key to use to determine what value to return in the model */
  modelKey: {
    type: String as PropType<"id" | "render">,
    default: "id",
  },

  /** The pages to show in the tab nav */
  options: {
    type: Array as PropType<selectOption[]>,
    required: true,
  },

  /** Allow multiple selections */
  multiple: {
    type: Boolean,
    required: false,
    default: false,
  },

  /** The key to use for what to display */
  displayKey: {
    type: String as PropType<"id" | "render">,
    default: "render",
  },
});

const { normalisedOptions, getLabel, isOptionSelected, updateModelValue } =
  useMultiselect(props, emit);
</script>
<template>
  <nav aria-label="Tab Navigation" class="tab-nav">
    <ul>
      <li v-for="page in normalisedOptions" :key="page[modelKey]">
        <base-button
          @click="updateModelValue(page[modelKey])"
          :aria-label="`Go to page ${page[modelKey]}`"
          :data-id="page[modelKey]"
          :class="{ active: isOptionSelected(page) }"
          :disabled="page.disabled"
        >
          {{ getLabel(page) }}
        </base-button>
      </li>
    </ul>
  </nav>
</template>
@/stories/Composables/useMultiselect
