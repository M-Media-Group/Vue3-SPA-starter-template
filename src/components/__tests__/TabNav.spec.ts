import { describe, expect, it } from "vitest";

import { RouterLinkStub, mount } from "@vue/test-utils";
import TabNav from "../TabNav.vue";
import "html-validate/vitest";

describe("Tab navigation", () => {
  it("renders correctly by default", () => {
    const wrapper = mount(TabNav, {
      props: {
        pages: [
          { render: "Page 1", id: "1" },
          { render: "Page 2", id: "2" },
          { render: "Page 3", id: "3" },
          { render: "Page 4", id: "4" },
          { render: "Page 5", id: "5" },
        ],
        modelValue: [],
      },
      global: {
        stubs: {
          RouterLink: RouterLinkStub,
        },
      },
    });

    const tabNav = wrapper.find("nav");
    expect(tabNav.exists()).toBe(true);

    const buttons = wrapper.findAll("button");
    expect(buttons.length).toBe(5);

    // Each text should be visible
    let i = 1;
    for (const button of buttons) {
      expect(button.text()).toBe("Page " + i++);
      //   No button should have an active class
      expect(button.classes("active")).toBe(false);
    }

    // @ts-ignore
    expect(wrapper.html()).toHTMLValidate();

    expect(wrapper.html()).toMatchSnapshot();
  });
});
