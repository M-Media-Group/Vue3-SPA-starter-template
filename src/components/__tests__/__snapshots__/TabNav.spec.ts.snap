// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tab navigation > renders correctly by default 1`] = `
"<nav aria-label="Tab Navigation" class="tab-nav">
  <ul>
    <li><button type="button" aria-label="Go to page 1" data-id="1" class="">Page 1
        <!--v-if-->
      </button></li>
    <li><button type="button" aria-label="Go to page 2" data-id="2" class="">Page 2
        <!--v-if-->
      </button></li>
    <li><button type="button" aria-label="Go to page 3" data-id="3" class="">Page 3
        <!--v-if-->
      </button></li>
    <li><button type="button" aria-label="Go to page 4" data-id="4" class="">Page 4
        <!--v-if-->
      </button></li>
    <li><button type="button" aria-label="Go to page 5" data-id="5" class="">Page 5
        <!--v-if-->
      </button></li>
  </ul>
</nav>"
`;
